File: app/python/src/grokking_fp/ch01_intro.py
"""第1章: 関数型プログラミング入門

関数型プログラミングの基本概念を Python で学びます。
命令型プログラミングとの違いを理解し、純粋関数の利点を実感することが目標です。
"""

from typing import Final


# =============================================================================
# 1.1 命令型 vs 関数型
# =============================================================================


def calculate_score_imperative(word: str) -> int:
    """命令型でワードスコアを計算する。

    命令型プログラミングは「どうやるか（HOW）」を記述します。

    Examples:
        >>> calculate_score_imperative("hello")
        5
        >>> calculate_score_imperative("")
        0
    """
    score = 0
    for _ in word:
        score += 1
    return score


def word_score(word: str) -> int:
    """関数型でワードスコアを計算する。

    関数型プログラミングは「何をするか（WHAT）」を記述します。

    Examples:
        >>> word_score("hello")
        5
        >>> word_score("Scala")
        5
        >>> word_score("")
        0
    """
    return len(word)


# =============================================================================
# 1.2 基本的な関数定義
# =============================================================================


def increment(x: int) -> int:
    """値を1増加させる。

    Examples:
        >>> increment(5)
        6
        >>> increment(0)
        1
        >>> increment(-1)
        0
    """
    return x + 1


def get_first_character(s: str) -> str:
    """文字列の最初の文字を取得する。

    Examples:
        >>> get_first_character("hello")
        'h'
        >>> get_first_character("Python")
        'P'
    """
    return s[0]


def add(a: int, b: int) -> int:
    """2つの数値を加算する。

    Examples:
        >>> add(2, 3)
        5
        >>> add(-1, 1)
        0
    """
    return a + b


def double(x: int) -> int:
    """値を2倍にする。

    Examples:
        >>> double(5)
        10
        >>> double(0)
        0
        >>> double(-3)
        -6
    """
    return x * 2


def greet(name: str) -> str:
    """挨拶メッセージを生成する。

    Examples:
        >>> greet("World")
        'Hello, World!'
        >>> greet("Python")
        'Hello, Python!'
    """
    return f"Hello, {name}!"


def to_uppercase(s: str) -> str:
    """文字列を大文字に変換する。

    Examples:
        >>> to_uppercase("hello")
        'HELLO'
        >>> to_uppercase("Python")
        'PYTHON'
    """
    return s.upper()


def reverse_string(s: str) -> str:
    """文字列を反転する。

    Examples:
        >>> reverse_string("hello")
        'olleh'
        >>> reverse_string("Python")
        'nohtyP'
    """
    return s[::-1]


# =============================================================================
# 1.3 条件分岐を含む関数
# =============================================================================


def absolute(x: int) -> int:
    """絶対値を返す。

    Examples:
        >>> absolute(5)
        5
        >>> absolute(-5)
        5
        >>> absolute(0)
        0
    """
    return x if x >= 0 else -x


def max_value(a: int, b: int) -> int:
    """2つの値のうち大きい方を返す。

    Examples:
        >>> max_value(3, 5)
        5
        >>> max_value(5, 3)
        5
        >>> max_value(3, 3)
        3
    """
    return a if a >= b else b


def min_value(a: int, b: int) -> int:
    """2つの値のうち小さい方を返す。

    Examples:
        >>> min_value(3, 5)
        3
        >>> min_value(5, 3)
        3
        >>> min_value(3, 3)
        3
    """
    return a if a <= b else b


def clamp(value: int, min_val: int, max_val: int) -> int:
    """値を指定された範囲内に収める。

    Examples:
        >>> clamp(5, 0, 10)
        5
        >>> clamp(-5, 0, 10)
        0
        >>> clamp(15, 0, 10)
        10
    """
    return max_value(min_val, min_value(value, max_val))


# =============================================================================
# 1.4 述語関数（真偽値を返す関数）
# =============================================================================


def is_even(n: int) -> bool:
    """偶数かどうかを判定する。

    Examples:
        >>> is_even(4)
        True
        >>> is_even(3)
        False
        >>> is_even(0)
        True
        >>> is_even(-2)
        True
    """
    return n % 2 == 0


def is_positive(n: int) -> bool:
    """正の数かどうかを判定する。

    Examples:
        >>> is_positive(5)
        True
        >>> is_positive(0)
        False
        >>> is_positive(-3)
        False
    """
    return n > 0


def is_empty(s: str) -> bool:
    """文字列が空かどうかを判定する。

    Examples:
        >>> is_empty("")
        True
        >>> is_empty("hello")
        False
    """
    return len(s) == 0


# =============================================================================
# 1.5 定数と型ヒント
# =============================================================================

# Final を使用した定数定義
MAX_SCORE: Final[int] = 100
MIN_SCORE: Final[int] = 0
DEFAULT_GREETING: Final[str] = "Hello"


def get_bounded_score(score: int) -> int:
    """スコアを有効な範囲内に収める。

    Examples:
        >>> get_bounded_score(50)
        50
        >>> get_bounded_score(150)
        100
        >>> get_bounded_score(-10)
        0
    """
    return clamp(score, MIN_SCORE, MAX_SCORE)

File: app/rust/src/ch01_intro.rs
//! 第1章: 関数型プログラミング入門
//!
//! 命令型プログラミングと関数型プログラミングの違いを学び、
//! Rust における関数定義の基本を理解します。

// =============================================================================
// 1.1 命令型 vs 関数型
// =============================================================================

/// 命令型スタイル: ステップバイステップで「どうやるか」を記述
///
/// # Examples
///
/// ```
/// use grokking_fp::ch01_intro::calculate_score_imperative;
/// assert_eq!(calculate_score_imperative("hello"), 5);
/// ```
pub fn calculate_score_imperative(word: &str) -> usize {
    let mut score = 0;
    for _ in word.chars() {
        score += 1;
    }
    score
}

/// 関数型スタイル: 「何をするか」を宣言的に記述
///
/// # Examples
///
/// ```
/// use grokking_fp::ch01_intro::word_score;
/// assert_eq!(word_score("hello"), 5);
/// assert_eq!(word_score("Rust"), 4);
/// assert_eq!(word_score(""), 0);
/// ```
pub fn word_score(word: &str) -> usize {
    word.len()
}

// =============================================================================
// 1.2 Rust の基本的な関数
// =============================================================================

/// 整数をインクリメント
///
/// # Examples
///
/// ```
/// use grokking_fp::ch01_intro::increment;
/// assert_eq!(increment(0), 1);
/// assert_eq!(increment(5), 6);
/// assert_eq!(increment(-1), 0);
/// ```
pub fn increment(x: i32) -> i32 {
    x + 1
}

/// 文字列の最初の文字を取得
///
/// # Examples
///
/// ```
/// use grokking_fp::ch01_intro::get_first_character;
/// assert_eq!(get_first_character("hello"), Some('h'));
/// assert_eq!(get_first_character("Rust"), Some('R'));
/// assert_eq!(get_first_character(""), None);
/// ```
pub fn get_first_character(s: &str) -> Option<char> {
    s.chars().next()
}

/// 2つの整数を加算
///
/// # Examples
///
/// ```
/// use grokking_fp::ch01_intro::add;
/// assert_eq!(add(2, 3), 5);
/// assert_eq!(add(-1, 1), 0);
/// assert_eq!(add(0, 0), 0);
/// ```
pub fn add(a: i32, b: i32) -> i32 {
    a + b
}

/// 整数を2倍にする
///
/// # Examples
///
/// ```
/// use grokking_fp::ch01_intro::double;
/// assert_eq!(double(5), 10);
/// assert_eq!(double(0), 0);
/// assert_eq!(double(-3), -6);
/// ```
pub fn double(x: i32) -> i32 {
    x * 2
}

// =============================================================================
// 1.3 文字列操作
// =============================================================================

/// 挨拶文を生成
///
/// # Examples
///
/// ```
/// use grokking_fp::ch01_intro::greet;
/// assert_eq!(greet("World"), "Hello, World!");
/// assert_eq!(greet("Rust"), "Hello, Rust!");
/// ```
pub fn greet(name: &str) -> String {
    format!("Hello, {}!", name)
}

/// 文字列を大文字に変換
///
/// # Examples
///
/// ```
/// use grokking_fp::ch01_intro::to_uppercase;
/// assert_eq!(to_uppercase("hello"), "HELLO");
/// assert_eq!(to_uppercase("Rust"), "RUST");
/// ```
pub fn to_uppercase(s: &str) -> String {
    s.to_uppercase()
}

/// 文字列を逆順にする
///
/// # Examples
///
/// ```
/// use grokking_fp::ch01_intro::reverse_string;
/// assert_eq!(reverse_string("hello"), "olleh");
/// assert_eq!(reverse_string("Rust"), "tsuR");
/// assert_eq!(reverse_string(""), "");
/// ```
pub fn reverse_string(s: &str) -> String {
    s.chars().rev().collect()
}

// =============================================================================
// 1.4 数値計算
// =============================================================================

/// 絶対値を計算
///
/// # Examples
///
/// ```
/// use grokking_fp::ch01_intro::absolute;
/// assert_eq!(absolute(5), 5);
/// assert_eq!(absolute(-5), 5);
/// assert_eq!(absolute(0), 0);
/// ```
pub fn absolute(x: i32) -> i32 {
    x.abs()
}

/// 最大値を取得
///
/// # Examples
///
/// ```
/// use grokking_fp::ch01_intro::max;
/// assert_eq!(max(3, 5), 5);
/// assert_eq!(max(10, 2), 10);
/// assert_eq!(max(-1, -5), -1);
/// ```
pub fn max(a: i32, b: i32) -> i32 {
    if a > b { a } else { b }
}

/// 最小値を取得
///
/// # Examples
///
/// ```
/// use grokking_fp::ch01_intro::min;
/// assert_eq!(min(3, 5), 3);
/// assert_eq!(min(10, 2), 2);
/// assert_eq!(min(-1, -5), -5);
/// ```
pub fn min(a: i32, b: i32) -> i32 {
    if a < b { a } else { b }
}

/// 値をクランプ（範囲内に制限）
///
/// # Examples
///
/// ```
/// use grokking_fp::ch01_intro::clamp;
/// assert_eq!(clamp(5, 0, 10), 5);
/// assert_eq!(clamp(-5, 0, 10), 0);
/// assert_eq!(clamp(15, 0, 10), 10);
/// ```
pub fn clamp(value: i32, min_val: i32, max_val: i32) -> i32 {
    if value < min_val {
        min_val
    } else if value > max_val {
        max_val
    } else {
        value
    }
}

// =============================================================================
// 1.5 述語関数（Predicate）
// =============================================================================

/// 偶数かどうかを判定
///
/// # Examples
///
/// ```
/// use grokking_fp::ch01_intro::is_even;
/// assert_eq!(is_even(2), true);
/// assert_eq!(is_even(3), false);
/// assert_eq!(is_even(0), true);
/// assert_eq!(is_even(-4), true);
/// ```
pub fn is_even(n: i32) -> bool {
    n % 2 == 0
}

/// 正の数かどうかを判定
///
/// # Examples
///
/// ```
/// use grokking_fp::ch01_intro::is_positive;
/// assert_eq!(is_positive(5), true);
/// assert_eq!(is_positive(-3), false);
/// assert_eq!(is_positive(0), false);
/// ```
pub fn is_positive(n: i32) -> bool {
    n > 0
}

/// 文字列が空かどうかを判定
///
/// # Examples
///
/// ```
/// use grokking_fp::ch01_intro::is_empty;
/// assert_eq!(is_empty(""), true);
/// assert_eq!(is_empty("hello"), false);
/// ```
pub fn is_empty(s: &str) -> bool {
    s.is_empty()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_word_score() {
        assert_eq!(word_score("hello"), 5);
        assert_eq!(word_score(""), 0);
        assert_eq!(word_score("Rust"), 4);
    }

    #[test]
    fn test_increment() {
        assert_eq!(increment(0), 1);
        assert_eq!(increment(-1), 0);
        assert_eq!(increment(i32::MAX - 1), i32::MAX);
    }

    #[test]
    fn test_get_first_character() {
        assert_eq!(get_first_character("hello"), Some('h'));
        assert_eq!(get_first_character(""), None);
    }

    #[test]
    fn test_greet() {
        assert_eq!(greet("World"), "Hello, World!");
    }

    #[test]
    fn test_is_even() {
        assert!(is_even(0));
        assert!(is_even(2));
        assert!(is_even(-4));
        assert!(!is_even(1));
        assert!(!is_even(-3));
    }

    #[test]
    fn test_clamp() {
        assert_eq!(clamp(5, 0, 10), 5);
        assert_eq!(clamp(-5, 0, 10), 0);
        assert_eq!(clamp(15, 0, 10), 10);
        assert_eq!(clamp(0, 0, 10), 0);
        assert_eq!(clamp(10, 0, 10), 10);
    }
}
